quantized tensor: add support for advanced indexing

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49129 quantized tensor: add support for advanced indexing**

Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

If helpful for reviewers, the things originally broken were, in order:
1. `computeDeviceType` did not handle `DispatchKey::QuantizedCPU` (fix: added)
2. quantization params were not present in `TensorIterator::set_output` so they can be used to properly create the quantized tensor (fix: created `TensorQuantizationOptions` and threaded it through the relevant places)
3. `index` kernel was not enabled for quantized dtypes (fix: enable it)

Note: this PR only handles per-Tensor qparams.  We don't expect to need this for per-channel qparams any time soon, ideally we can pay the eng cost for enabling that when it is needed.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25451651](https://our.internmc.facebook.com/intern/diff/D25451651)

vkuzo committed:
[wip] quantized tensor: add support for advanced indexing

Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

At least a few things need to happen before this being ready for review:
1. Verify if the general idea of `TensorAdditionalOptions` is the right abstraction
2. Verify if `TensorAdditionalOptions` needs the same style as `TensorOptions` (no optionals, etc)

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] quantized tensor: add support for advanced indexing"


Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

At least a few things need to happen before this being ready for review:
1. Verify if the general idea of `TensorAdditionalOptions` is the right abstraction
2. Verify if `TensorAdditionalOptions` needs the same style as `TensorOptions` (no optionals, etc)

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25451651](https://our.internmc.facebook.com/intern/diff/D25451651)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -475,17 +475,38 @@ void TensorIteratorBase::allocate_or_resize_outputs() {
+      if (options.dtype() == at::kQInt8 || options.dtype() == at::kQUInt8 ||
+          options.dtype() == at::kQInt32) {
+        // get the first input and copy its quantization parameters
+        const auto& first_input = operands_[num_outputs_];
```
So what if the quantization parameters on the two tensors are different? That seems like a totally reasonable situation, and then copying the quantization parameters from the first input seems like a totally arbitrary choice.

ezyang replied:
If the correct quantization of the output tensor is something that varies from operator to operator, it may make sense to force kernels to do the computation of the quantization. There is some precedent for doing this in TensorIterator with the `static_shape` parameter, where kernels can say "Hey, I know you've got some built-in shape computation, but I actually happen to know what the output shape is, please use this." I could easily imagine something similar for quantization.

vkuzo replied:
yeah, I don't think the current solution is super clean, I'm also not sure if the alternative would be better.

The two high level ways I saw to do this were:
1. add a quantized path to the current fp32 indexing logic (current PR).  This reuses the size calculations and the kernels, but has to add some metadata for quantized tensors in places where it previously did not exist.
2. dispatch to something like `quantized_index`, which would have to reimplement the indexing logic (with some pieces potentially reusable from fp32, but likely not all)

Thoughts on any options I might be missing here?  Does 1 sound like the right call?

vkuzo replied:
for now, it's option 1, with the non-relevant parts asserted out, but I'm flexible

ezyang replied:
@vkuzo The nuance here is that the code paths you are editing in TensorIterator aren't just for fp32 indexing: they are for *every unary, binary and reduction operation in PyTorch*. So if you want to add something here, there is an obligation to at least have some plausible argument why the logic here is appropriate for all other operations.

I don't mind option one, but this PR ain't it. An implementation that more closely follows the idea of (1) is to take the existing indexing kernel (not in TensorIterator) and program TensorIterator on what the intended quantization result is.

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1281,12 +1302,25 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
+          if (options.dtype() == at::kQInt8 || options.dtype() == at::kQUInt8 ||
+              options.dtype() == at::kQInt32) {
+            // quantized path
+            TORCH_INTERNAL_ASSERT(quant_options.has_value());
```
Wouldn't it be faster to test if `quant_options.has_value()` rather than having to do three conditions on dtype.

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1281,12 +1302,25 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
+          }
       } else {
           op.tensor = at::empty_strided(sizes, strides, options);
       }
```
Below, there is logic for handling what happens in the `out=` case. Imagine that you do indexing into a pre-existing quantized tensor. What should the quantization parameters be? What if the tensor is to be *resized* in this situation?

vkuzo replied:
asserted it out for now

ezyang commented:
The change to MetaBase::set_output looks incomplete. Oddly, the CI is passing, this is probably because none of the quantized operations have been made structured yet.

vkuzo committed:
Update on "quantized tensor: add support for advanced indexing"


Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

If helpful for reviewers, the things originally broken were, in order:
1. `computeDeviceType` did not handle `DispatchKey::QuantizedCPU` (fix: added)
2. quantization params were not present in `TensorIterator::set_output` so they can be used to properly create the quantized tensor (fix: created `TensorQuantizationOptions` and threaded it through the relevant places)
3. `index` kernel was not enabled for quantized dtypes (fix: enable it)

Note: this PR only handles per-Tensor qparams.  We don't expect to need this for per-channel qparams any time soon, ideally we can pay the eng cost for enabling that when it is needed.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25451651](https://our.internmc.facebook.com/intern/diff/D25451651)

[ghstack-poisoned]

vkuzo commented:
> The change to MetaBase::set_output looks incomplete. Oddly, the CI is passing, this is probably because none of the quantized operations have been made structured yet.

to clarify, do you mean accounting for the changes in https://github.com/pytorch/pytorch/pull/48718?  Yes, that was not in the original PR, but is there in the latest update.

vkuzo committed:
clang format on "quantized tensor: add support for advanced indexing"


Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

If helpful for reviewers, the things originally broken were, in order:
1. `computeDeviceType` did not handle `DispatchKey::QuantizedCPU` (fix: added)
2. quantization params were not present in `TensorIterator::set_output` so they can be used to properly create the quantized tensor (fix: created `TensorQuantizationOptions` and threaded it through the relevant places)
3. `index` kernel was not enabled for quantized dtypes (fix: enable it)

Note: this PR only handles per-Tensor qparams.  We don't expect to need this for per-channel qparams any time soon, ideally we can pay the eng cost for enabling that when it is needed.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25451651](https://our.internmc.facebook.com/intern/diff/D25451651)

[ghstack-poisoned]

vkuzo committed:
not for land, just testing on "quantized tensor: add support for advanced indexing"


Summary:

Implements support for the indexing of quantized tensors with lists of
dims, such as

```
xq_slice = xq[:, [0], :, :]
```

If helpful for reviewers, the things originally broken were, in order:
1. `computeDeviceType` did not handle `DispatchKey::QuantizedCPU` (fix: added)
2. quantization params were not present in `TensorIterator::set_output` so they can be used to properly create the quantized tensor (fix: created `TensorQuantizationOptions` and threaded it through the relevant places)
3. `index` kernel was not enabled for quantized dtypes (fix: enable it)

Note: this PR only handles per-Tensor qparams.  We don't expect to need this for per-channel qparams any time soon, ideally we can pay the eng cost for enabling that when it is needed.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_advanced_indexing
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25451651](https://our.internmc.facebook.com/intern/diff/D25451651)

[ghstack-poisoned]

vkuzo commented:
thanks for the review!  We went with a simpler and less performant approach of using the fp32 kernel for now in https://github.com/pytorch/pytorch/pull/49346, we might revisit supporting this natively in the future.
